packages/model-viewer/src/features/scene-graph/model.ts

```diff
@@ -13,42 +13,275 @@
+    super(mesh.name);
+    this[$mesh] = mesh;
+    const {gltf, threeGLTF} = correlatedSceneGraph;
+
+    // TODO: Remove the associationKey 'work arounds' after fixing Three.js
+    // associations. This is needed for now because Three.js does not create
+    // associations with SkinnedMeshes (glTF primitives of an animated object)
+    // and incorrect associations are formed when a mesh has multiple
+    // primitives.
+    let associationKey: Object3D = mesh;
+    // Work around 1, skinned meshes (glTF primitives) have no association but
+    // the parent (mesh) does, which maps to a 'node'.
+    if (mesh instanceof SkinnedMesh && mesh.parent != null) {
+      associationKey = mesh.parent;
+    }
+    let gltfMeshReference = correlatedSceneGraph.threeObjectMap.get(
+                                associationKey) as GLTFReference;
+    // Work around 2, the association value is a material not a node.
+    if (gltfMeshReference.type === 'materials') {
+      associationKey = mesh.parent!;
+      gltfMeshReference = correlatedSceneGraph.threeObjectMap.get(
+                              associationKey) as GLTFReference;
+    }
+
+    const {type: nodes, index: nodeIndex} = gltfMeshReference;
+
+    // Should have the correct reference type now.
+    if (nodes !== 'nodes') {
+      console.error('Expected type \'nodes\' but got ' + nodes);
+    }
+    // Gets the mesh index from the node.
+    const meshIndex = ((gltf[nodes] || []) as GLTFNode[])[nodeIndex].mesh!;
+    // The gltf mesh array to sample from.
+    const meshElementArray = gltf['meshes'] || [];
+    // List of primitives under the mesh.
+    const gltfPrimitives =
+        (meshElementArray[meshIndex].primitives || []) as Primitive[];
+
+    for (const primitive of gltfPrimitives) {
+      // Maps the primitive default to a material.
+      this[$materials].set(
+          primitive.material!, mvMaterials[primitive.material!]);
+
+      if (primitive.extensions &&
+          primitive.extensions['KHR_materials_variants']) {
+        const variantsExtension =
+            primitive.extensions['KHR_materials_variants'] as
+            KHRMaterialsVariants;
+        const extensions = threeGLTF.parser.json.extensions;
+        const variantNames = extensions['KHR_materials_variants'].variants;
+        // Provides definition now that we know there are variants to support.
+        this[$variantInfo] = new Map<string, Material>();
+        for (const mapping of variantsExtension.mappings!) {
+          // Maps variant indices to Materials.
+          this[$materials].set(mapping.material, mvMaterials[mapping.material]);
+          for (const variant of mapping.variants) {
+            const {name} = variantNames[variant];
+            this[$variantInfo].set(name, mvMaterials[mapping.material]);
+          }
+        }
+      }
+    }
+  }
+
+  get mesh() {
+    return this[$mesh];
+  }
+
+  async setActiveMaterial(material: number|Material):
```

Comment:
Kind of an odd signature; maybe make this two separate functions?

Reply:
Yeah, actually I think passing a material is improper here as it could be any material from anywhere. It's good to be able to do that but it would bypass all the bookkeeping right now, that would need to be handled differently, going to kick the can down the road.